src/utils/cosmosdb_model_versioned.ts

```diff
@@ -0,0 +1,192 @@
+
+    // the ID of each document version is composed of the document ID and its version
+    // this makes it possible to detect conflicting updates (concurrent creation of
+    // profiles with the same profile ID and version)
+    const modelId = this.getModelId(o);
+    const versionedModelId = generateVersionedModelId(modelId, version);
+
+    const newDocument = {
+      ...o,
+      id: versionedModelId,
+      version
+    } as TN & VersionedModel & BaseModel;
+
+    return super.create(newDocument, options);
+  };
+
+  /**
+   * Creates a new version from a full item definition
+   */
+  public upsert = (o: TN): TaskEither<CosmosErrors, TR> => {
+    // if we get an explicit version number from the new document we use that,
+    // or else we get the last version by querying the database
+    const currentVersion: NonNegativeInteger | undefined = o.version;
+
+    const modelId = this.getModelId(o);
+    return (currentVersion === undefined
+      ? this.getLastVersion(modelId)
+      : fromEither<CosmosErrors, NonNegativeInteger>(right(currentVersion))
+    ).chain(nextVersion =>
+      super.create({
+        ...o,
+        id: generateVersionedModelId(modelId, nextVersion),
+        version: nextVersion
+      } as TN & VersionedModel & BaseModel)
+    );
+  };
+
+  /**
+   *  Find the last version of a document.
+   *
+   *  Pass the partitionKey field / values if it differs from the modelId
+   *  to avoid multi-partition queries.
+   */
+  public findLastVersionByModelId(
+    modelId: string,
+    partitionKey?: string
+  ): TaskEither<CosmosErrors, Option<TR>> {
+    const q: SqlQuerySpec = {
+      parameters: [
+        {
+          name: "@modelId",
+          value: modelId
+        }
+      ],
+      // Note: do not use ${collectionName} here as it may contain special characters
+      query: `SELECT TOP 1 * FROM m WHERE m.${this.modelIdKey} = @modelId ORDER BY m.version DESC`
+    };
+    const queryItems = this.container.items.query(q, {
+      maxItemCount: 1,
+      partitionKey: partitionKey !== undefined ? partitionKey : modelId
+    }).fetchAll;
+    return tryCatch<CosmosErrors, PromiseType<ReturnType<typeof queryItems>>>(
+      queryItems,
+      _ => CosmosErrorResponse(_ as ErrorResponse)
+    )
+      .map(_ => fromNullable(_.resources[0]))
+      .chain(_ =>
+        _.isSome()
+          ? fromEither(
+              this.retrievedItemT
+                .decode(_.value)
+                .map(some)
+                .mapLeft(CosmosDecodingError)
+            )
+          : fromEither(right(none))
+      );
+  }
+
+  /**
+   * Returns the value of the model ID for the provided item
+   */
+  protected getModelId = (o: T) =>
+    ModelId.decode(String(o[this.modelIdKey])).value as ModelId;
+
+  /**
+   * Returns the most recent version of the model from the database or 0 if
+   * no previous version is found.
+   */
+  private getLastVersion = (modelId: ModelId) =>
```

Comment:
the name is a little bit misleading since this not get the latest version but the next one (which is called "current" above)

this is hard to grasp: getLastVersion = findLastVersionByModelId + 1

Reply:
ops you are right, it should be called `getNextVersion`